save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/utils/ResultsChecker.kt

```diff
@@ -0,0 +1,145 @@
+        // Expected warnings that were found among of Actual warnings
+        val expectedWarningsMatchedWithActual = expectedWarnings
+            .matchWithActualWarnings(actualWarnings, actualMatchedWithExpectedWarnings)
+
+        val missingWarnings = expectedWarnings - expectedWarningsMatchedWithActual
+        val unexpectedWarnings = actualWarnings - actualMatchedWithExpectedWarnings
+
+        return when (missingWarnings.isEmpty() to unexpectedWarnings.isEmpty()) {
+            false to true -> createFailFromSingleMiss(EXPECTED_BUT_NOT_RECEIVED, missingWarnings)
+            false to false -> createFailFromDoubleMiss(missingWarnings, unexpectedWarnings)
+            true to true -> Pass(null)
+            true to false -> if (warnPluginConfig.exactWarningsMatch == false) {
+                Pass(
+                    "$UNEXPECTED: $unexpectedWarnings",
+                    "$UNEXPECTED: ${unexpectedWarnings.size} warnings"
+                )
+            } else {
+                createFailFromSingleMiss(UNEXPECTED, unexpectedWarnings)
+            }
+            else -> Fail("N/A", "N/A")
+        }
+    }
+
+    private fun List<Warning>.matchWithActualWarnings(
+        actualWarnings: List<Warning>,
+        actualMatchedWithExpectedWarnings: MutableList<Warning>
+    ) = this.filter { expected ->
+        val matchedWarning = actualWarnings.find { actual ->
+            // matched line and column
+            (expected.line == actual.line && expected.column == actual.column) &&
+                    // matched text of the message
+                    expected.message.createRegexFromMessage().matches(actual.message)
+        }
+
+        matchedWarning?.let {
+            actualMatchedWithExpectedWarnings.add(matchedWarning)
+        }
+        matchedWarning != null
+    }
+
+    private fun createFailFromSingleMiss(baseText: String, warnings: List<Warning>) =
+            Fail("$baseText: $warnings", "$baseText (${warnings.size})")
+
+    private fun createFailFromDoubleMiss(missingWarnings: List<Warning>, unexpectedWarnings: List<Warning>) =
+            Fail(
+                "$EXPECTED_BUT_NOT_RECEIVED: $missingWarnings, and ${UNEXPECTED.lowercase()}: $unexpectedWarnings",
+                "$EXPECTED_BUT_NOT_RECEIVED (${missingWarnings.size}), and ${UNEXPECTED.lowercase()} (${unexpectedWarnings.size})"
+            )
+
+    private fun String.createRegexFromMessage(): Regex {
+        // patternForRegexInWarning cannot be null, as it has a default value
+        val openingDelimiter = warnPluginConfig.patternForRegexInWarning!![0]
+        val closingDelimiter = warnPluginConfig.patternForRegexInWarning[1]
+
+        // searching all delimited regex in the warning
+        val foundSubStringsWithRegex = this
+            .findDelimitedSubStringsWith(openingDelimiter, closingDelimiter)
+            .entries
+            .sortedBy { it.key }
+
+        val resultWithRegex = foundSubStringsWithRegex.mapIndexed { i, entry ->
```

Comment:
Is this logic equivalent to something like
```kotlin
foundSubStringsWithRegex.foldRight(this) { acc, entry ->
            acc.removeRange(entry.value, entry.value + closingDelimiter.length)
                .removeRange(entry.key, entry.key + closingDelimiter.length)
}
```
?

Reply:
discussed offline, the main aim of my logic is to escape special symbols